File: 2022/day2/c++/part2.cpp
#include <fstream>
#include <iostream>
#include <string>

using namespace std;

enum Shapes { ROCK = 'A', PAPER = 'B', SCISSORS = 'C' };

int get_shape_score(const char move) {
  return move == Shapes::ROCK ? 1 : move == Shapes::PAPER ? 2 : 3;
}

int get_score(const char move, const char outcome) {
  switch (outcome) {
    case 'X':  // Lose
      return get_shape_score(move == Shapes::ROCK    ? Shapes::SCISSORS
                             : move == Shapes::PAPER ? Shapes::ROCK
                                                     : Shapes::PAPER);
    case 'Y':  // Draw
      return get_shape_score(move) + 3;
    case 'Z':  // Win
      return get_shape_score(move == Shapes::ROCK    ? Shapes::PAPER
                             : move == Shapes::PAPER ? Shapes::SCISSORS
                                                     : Shapes::ROCK) +
             6;
    default:  // Unknown
      return 0;
  }
}

int main() {
  ifstream input("../input.txt");

  if (!input.is_open()) {
    cerr << "Unable to open input file\n";

    return 1;
  }

  size_t total_score = 0;

  string line;
  while (getline(input, line)) total_score += get_score(line.at(0), line.at(2));

  input.close();

  cout << total_score << '\n';
}

File: 2022/day2/javascript/part2.mjs
import { readFileSync } from 'node:fs';

const input = readFileSync('../input.txt', 'utf-8');

const rounds = input.split(/\r?\n/);

const Shapes = {
  ROCK: 'A',
  PAPER: 'B',
  SCISSORS: 'C'
};

function getShapeScore(move) {
  return move === Shapes.ROCK ? 1 : move === Shapes.PAPER ? 2 : 3;
}

function getScore(move, outcome) {
  switch (outcome) {
    case 'X': // Lose
      return getShapeScore(
        move === Shapes.ROCK
          ? Shapes.SCISSORS
          : move === Shapes.PAPER
          ? Shapes.ROCK
          : Shapes.PAPER
      );
    case 'Y': // Draw
      return getShapeScore(move) + 3;
    case 'Z': // Win
      return (
        getShapeScore(
          move === Shapes.ROCK
            ? Shapes.PAPER
            : move === Shapes.PAPER
            ? Shapes.SCISSORS
            : Shapes.ROCK
        ) + 6
      );
    default: // Unknown
      return 0;
  }
}

let totalScore = 0;

for (const round of rounds) {
  if (round === '') break;

  const [opponentMove, currentMove] = round.split(' ');

  totalScore += getScore(opponentMove, currentMove);
}

console.log(totalScore);
